env/context.go

```diff
@@ -5,59 +5,174 @@
+func NewContextFromGlobalContext(g *GlobalContext) *KBFSContext {
+	g.Init()
+	g.ConfigureConfig()
+	g.ConfigureLogging()
+	g.ConfigureCaches()
+	g.ConfigureMerkleClient()
+	c := &KBFSContext{g: g}
+	c.initKBFSSocket()
+	return c
+}
 
 // NewContext constructs a context
-func NewContext() *Context {
-	// TODO: Remove direct use of libkb.G
-	libkbOnce.Do(func() {
-		libkb.G.Init()
-		libkb.G.ConfigureConfig()
-		libkb.G.ConfigureLogging()
-		libkb.G.ConfigureCaches()
-		libkb.G.ConfigureMerkleClient()
-	})
-	return &Context{g: libkb.G}
+func NewContext() *KBFSContext {
+	return NewContextFromGlobalContext(libkb.G)
 }
 
 // GetLogDir returns log dir
-func (c Context) GetLogDir() string {
+func (c *KBFSContext) GetLogDir() string {
 	return c.g.Env.GetLogDir()
 }
 
 // GetDataDir returns log dir
-func (c Context) GetDataDir() string {
+func (c *KBFSContext) GetDataDir() string {
 	return c.g.Env.GetDataDir()
 }
 
 // GetRunMode returns run mode
-func (c Context) GetRunMode() libkb.RunMode {
+func (c *KBFSContext) GetRunMode() libkb.RunMode {
 	return c.g.GetRunMode()
 }
 
 // GetSocket returns a socket
-func (c Context) GetSocket(clearError bool) (net.Conn, rpc.Transporter, bool, error) {
+func (c *KBFSContext) GetSocket(clearError bool) (
+	net.Conn, rpc.Transporter, bool, error) {
 	return c.g.GetSocket(clearError)
 }
 
 // ConfigureSocketInfo configures a socket
-func (c Context) ConfigureSocketInfo() error {
+func (c *KBFSContext) ConfigureSocketInfo() error {
 	return c.g.ConfigureSocketInfo()
 }
 
 // NewRPCLogFactory constructs an RPC logger
-func (c Context) NewRPCLogFactory() *libkb.RPCLogFactory {
+func (c *KBFSContext) NewRPCLogFactory() *libkb.RPCLogFactory {
 	return &libkb.RPCLogFactory{Contextified: libkb.NewContextified(c.g)}
 }
+
+func (c *KBFSContext) getSandboxSocketFile() string {
+	sandboxDir := c.g.Env.HomeFinder.SandboxCacheDir()
+	if sandboxDir == "" {
+		return ""
+	}
+	return filepath.Join(sandboxDir, kbfsSocketFile)
+}
+
+func (c *KBFSContext) getkbfsSocketFile() string {
+	e := c.g.Env
+	return e.GetString(
+		func() string { return c.getSandboxSocketFile() },
+		// TODO: maybe add command-line option here
+		func() string { return os.Getenv("KBFS_SOCKET_FILE") },
+		func() string { return filepath.Join(e.GetRuntimeDir(), kbfsSocketFile) },
+	)
+}
+
+func (c *KBFSContext) newTransportFromSocket(s net.Conn) rpc.Transporter {
+	return rpc.NewTransport(s, c.NewRPCLogFactory(), libkb.WrapError)
+}
+
+// GetKBFSSocket dials the socket configured in `c.kbfsSocket`.
+func (c *KBFSContext) GetKBFSSocket(clearError bool) (
+	net.Conn, rpc.Transporter, bool, error) {
+
+	var err error
+	c.g.Trace("GetSocket", func() error { return err })()
+
+	// Protect all global socket wrapper manipulation with a
+	// lock to prevent race conditions.
+	c.kbfsSocketMtx.Lock()
+	defer c.kbfsSocketMtx.Unlock()
+
+	needWrapper := false
+	if c.kbfsSocketWrapper == nil {
+		needWrapper = true
+		c.g.Log.Debug("| empty socket wrapper; need a new one")
```

Comment:
We don't usually use the "|" thing in KBFS logs, but if you have a reason for wanting to keep it consistent with the service logs, that's fine with me.